### DISCUSSION TITLE
SRT Push Publisher

### DISCUSSION THREAD

**Discussion by choigilhoon** [OvenMediaLabs/OvenMediaEngine#1299] [Announcement]
The SRT Push Publisher has been added.

OME's SRT push publisher operates in caller mode.

[Server.xml]
```xml
<Applications>
  <Application>
     ...
    <Publishers>
      ...
      <SRTPush>
      </SRTPush>
    </Publishers>
  </Application>
</Applications>
```
[API]
```
POST /v1/vhosts/{vhost}/apps/{app}:startPush

{
  "id": "{unique_push_id}",
  "stream": {
    "name": "{output_stream_name}",
    "variantNames": []
  },
  "protocol": "srt",
  "url": "srt://{host}[:port]?mode=caller&latency=120000&timeout=500000",
  "streamKey": ""
}
```

For more details, please refer to the manual:
https://airensoft.gitbook.io/ovenmediaengine/v/dev/push-publishing
https://airensoft.gitbook.io/ovenmediaengine/v/dev/rest-api/v1/virtualhost/application/push

You can test it with the latest master branch.
It will be included in the v0.15.14 release.
We appreciate your feedback.

**Comment by bchah** [comment]
Thank you @choigilhoon !! The SRT push works great but right now there is only code for starting the push, not stopping it or getting information from the API. Are you planning to add the rest of this soon or do you need a PR for it?

**Reply by choigilhoon** [reply]
Thank you for your prompt feedback!
I have identified that the response information was missing in the GET API, and it has been resolved in this commit: (https://github.com/AirenSoft/OvenMediaEngine/commit/c573683cf7ee82def0aa967dd8961557e50d958c).
The response information is provided in the following format:
```json
{
    "message": "OK",
    "response": [
        {
            "app": "app",
            "createdTime": "2023-07-11T13:06:46.636+09:00",
            "finishTime": "1970-01-01T09:00:00.000+09:00",
            "id": "unique_push_id",
            "protocol": "srt",
            "sentBytes": 1495087,
            "sentTime": 15289,
            "sequence": 0,
            "startTime": "2023-07-11T13:06:55.657+09:00",
            "state": "Pushing",
            "stream": {
                "name": "stream_test",
                "trackIds": [],
                "variantNames": []
            },
            "totalsentBytes": 1495087,
            "totalsentTime": 15289,
            "url": "srt://{host}:{port}?mode=caller",
            "vhost": "default"
        }
    ],
    "statusCode": 200
}
```

However, i was unable to reproduce the issue with the STOP API. I have observed that upon making a STOP API request, the stream on the SRT listener server (OME) is immediately deleted.

For reference, here is the information from my test:
[Request]
```
POST https://{host}:{port}/v1/vhosts/default/apps/app:stopPush

{
  "id": "unique_push_id"
}
```
[Response]
```
{
    "message": "[HTTP] Success (200)",
    "statusCode": 200
}
```

**Comment by bchah** [comment]
@choigilhoon thank you, you were right! The stopping issue was on my side, since we didn't get the API response from the start command it was not issuing the stop command 🤪 Testing with the master branch now and it works great.

I should also point out to the community that this feature also works with HEVC! So amazing, just a few weeks ago the idea of doing this was just a dream. But Airensoft continues to make our dreams come true 😇

Now we can use OME as a means of playout. You guys are the kings! 🥇

**Comment by alexthedamager** [comment]
To confirm - this is for sending an OME stream to a different SRT-supported server, and doesn't allow us to subscribe directly to an OME stream via SRT? Or am I mistaken?

**Reply by bchah** [reply]
OME has SRT listener mode (where 'callers' send streams). This adds caller mode to OME, so it can now send streams to devices or applications that support SRT listener mode.

But this isn't an "SRT publisher" for other devices to pull the stream - it adds "push out" to the current "push in" functionality. I'd back that feature request for sure though!

**Reply by alexthedamager** [reply]
Thanks for the thorough explanation! Caller mode is great, "SRT publisher" would for sure be amazing too and something I'd definitely make use of.

**Comment by tonywingmah** [comment]
Any plans for SRT playback in listener mode? I want to play SRT with OBS using home internet.
